File: 4. C++ Basics/4. Translation/Exercise_1/robot_localization_1.cpp
#include <iostream>
#include <vector>

using namespace std;
//Given the list motions=[1,1] which means the robot 
//moves right and then right again, compute the posterior 
//distribution if the robot first senses red, then moves 
//right one, then senses green, then moves right again, 
//starting with a uniform prior distribution.

vector <float> p (5 ,0.2);
vector <string> world = {"green", "red", "red", "green", "green"};
float pHit = 0.6;
float pMiss = 0.2;

vector <float> sense(vector <float> p, string Z);
vector <float> move(vector <float> p, int U);
int main()
{
    vector <string> measurements = {"red", "red"};
    vector <int> motions = {1,1};
    for (int i=0; i< measurements.size(); i++)
    {
        p = sense(p, measurements[i]);
        p = move(p, motions[i]);
    }
    for (float n: p) cout << n << " ";
    return 0;
}


vector <float> sense(vector <float> p, string Z)
{
    /*
    p = initial probability distribution list
    Z = each measurement
    q = new normalized distribution list
    */
    vector <float> q (p.size());
    for (int each_index = 0; each_index < p.size(); each_index++)
    {
        if (world[each_index] == Z)
        {
            q[each_index] = p[each_index]*pHit;
        }
        else
        {
            q[each_index] = p[each_index]*pMiss;
        }
    }
    float total = 0; for (float n: q) total += n; //https://stackoverflow.com/questions/3221812/how-to-sum-up-elements-of-a-c-vector
    for (float& n: q) n /= total;   //https://stackoverflow.com/questions/41030911/how-to-multiply-a-vector-and-scalar-in-c
    return q;
}

float pExact = 0.8;
float pOvershoot = 0.1;
float pUndershoot = 0.1;
vector <float> move(vector <float> p, int U)
{
    /*
    p = given probability distribution list
    U = amount of steps robot has moved to right (assumed as right for now)
    */
    //INEXACT MOTION
    vector <float> q (p.size());
    vector <float> temp = {pUndershoot, pExact, pOvershoot};
    for (int i = 0; i<p.size(); i++)
    {
        float undershoot_index    = (i + (U-1)) % p.size();
        float exact_index         = (i + (U)) % p.size();
        float overshoot_index     = (i + (U+1)) % p.size();

        //cout << i << undershoot_index << exact_index << overshoot_index << q[i] << endl;

        q[ undershoot_index ]   = p[i]*temp[0] + q[ undershoot_index ];
        q[ exact_index ]        = p[i]*temp[1] + q[ exact_index ];
        q[ overshoot_index ]    = p[i]*temp[2] + q[ overshoot_index ];
    }
    return q;
}

File: 4. C++ Basics/4. Translation/Exercise_1/robot_localization_1.py
#Given the list motions=[1,1] which means the robot 
#moves right and then right again, compute the posterior 
#distribution if the robot first senses red, then moves 
#right one, then senses green, then moves right again, 
#starting with a uniform prior distribution.

p=[0.2, 0.2, 0.2, 0.2, 0.2]
world=['green', 'red', 'red', 'green', 'green']
pHit = 0.6
pMiss = 0.2

def sense(p, Z):
    """
    p = initial probability distribution list
    Z = each measurement
    q = new normalized distribution list
    """
    q = []
    for each_index in range(len(p)):
        if (world[each_index] == Z):    #joint probability
            q.append(p[each_index]*pHit)
        else:
            q.append(p[each_index]*pMiss)
    total = sum(q)
    q = [ x/total for x in q]   #normalization
    return q

pExact = 0.8
pOvershoot = 0.1
pUndershoot = 0.1
def move(p, U):
    """
    p = given probability distribution list
    U = amount of steps robot has moved to right (assumed as right for now)
    """
    # INEXACT MOTION
    q = [ 0 for _ in p ]
    temp = [pUndershoot, pExact, pOvershoot]
    for i in range(len(p)):

        undershoot_index    = (i + (U-1)) % len(p)
        exact_index         = (i + (U)) % len(p)
        overshoot_index     = (i + (U+1)) % len(p)

        #print(i , undershoot_index , exact_index, overshoot_index, q)

        q[ undershoot_index ]   = p[i]*temp[0] + q[ undershoot_index ]
        q[ exact_index ]        = p[i]*temp[1] + q[ exact_index ]
        q[ overshoot_index ]    = p[i]*temp[2] + q[ overshoot_index ]
    return q

import matplotlib.pyplot as plt
def output_map(grid):
    """
    outputs a bar chart showing the probabilities of each grid space
    """
    x_labels = []
    for index in range(len(grid)):
        x_labels.append(index)
    plt.bar(x_labels, grid)
 
    # Create names on the axes
    plt.title('Probability of the robot being at each space on the grid')
    plt.xlabel('1D Space')
    plt.ylabel('Probability')

    # Show graphic
    plt.show(block=True)
#
# ADD CODE HERE
#
# for each_index in range(len(measurements)):
#     p = sense(p, measurements[each_index])
measurements = ['red', 'red']
motions = [1,1]

for i in range(len(measurements)):
    p = sense(p, measurements[i])
    p = move(p, motions[i])
        
print(p) 
output_map(p)      
